src/main/java/com/amazon/opendistroforelasticsearch/security/auth/RolesInjector.java

```diff
@@ -0,0 +1,104 @@
+
+package com.amazon.opendistroforelasticsearch.security.auth;
+
+import com.amazon.opendistroforelasticsearch.security.support.ConfigConstants;
+import com.amazon.opendistroforelasticsearch.security.user.User;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.transport.TransportAddress;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
+import java.net.InetAddress;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * This is used to inject opendistro-roles into the request when there is no user involved, like periodic plugin
+ * background jobs. The roles injection is done using thread-context at transport layer only. You can't inject
+ * roles using REST api. Using this we can enforce fine-grained-access-control for the transport layer calls plugins make.
+ *
+ * Format for the injected string: user_name|role_1,role_2
+ * User name is ignored. And roles are opendistro-roles.
+ */
+
+final public class RolesInjector {
+    protected final Logger log = LogManager.getLogger(RolesInjector.class);
+    private final boolean enabled;
+    private User user = null;
+    private Set<String> roles = null;
+    private final ThreadContext threadContext;
+
+    public RolesInjector(final Settings settings, final ThreadContext ctx) {
+        this.threadContext = ctx;
+        this.enabled = settings.getAsBoolean(ConfigConstants.OPENDISTRO_SECURITY_INJECTED_ROLES_ENABLED, true);
+
+        parseInjectedStr(ctx.getTransient(ConfigConstants.OPENDISTRO_SECURITY_INJECTED_ROLES));
+        if(log.isDebugEnabled()){
+            log.debug("Injected role str: "+ ctx.getTransient(ConfigConstants.OPENDISTRO_SECURITY_INJECTED_ROLES));
+        }
+        addRemoteAddr(threadContext);
+    }
+
+    private void addRemoteAddr(ThreadContext threadContext) {
+        if(threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS) != null)
+            return;
+
+        threadContext.putTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS,
+                new TransportAddress(InetAddress.getLoopbackAddress(), 9300));
+    }
+
+    /*
+     * Input string format: user|role_1,role2
+     */
+    private void parseInjectedStr(final String injectedStr) {
+        if (Strings.isNullOrEmpty(injectedStr))
+            return;
+
+        String[] strs = injectedStr.split("\\|");
+        if (strs.length == 0) {
+            log.error("Roles injected string malformed, could not extract parts. User string was '{}.'" +
+                    " Roles injection failed.", injectedStr);
+            return;
+        }
+        if (Strings.isNullOrEmpty(strs[0])) {
+            log.error("Username must not be null, injected string was '{}.' Roles injection failed.", injectedStr);
+            return;
+        }
+        this.user = new User(strs[0]);
```

Comment:
put `user` to thread context and remove `getUser()`

Reply:
Addressed.